File: leetcode/Min_Stack.cpp
class MinStack {
private:
    stack<int> st;
    stack<int> min_st;
public:
    void push(int x) {
        if(min_st.empty() || x <= min_st.top()) min_st.push(x);
        st.push(x);
    }

    void pop() {
        if(st.empty()) return;
        
        if(st.top() == min_st.top()) min_st.pop();
        st.pop();
    }

    int top() {
        if(!st.empty()) return st.top();
    }

    int getMin() {
        if(!min_st.empty()) return min_st.top();
    }
};

File: leetcode/java/MinStack.java
// MaxStack can be implemented in a similar fashion as well.

class MinStack {
    private final Deque<Integer> st;
    private final Deque<Integer> mst;

    /** initialize your data structure here. */
    public MinStack() {
        st = new ArrayDeque<>();
        mst = new ArrayDeque<>();
    }
    
    public void push(int x) {
        st.push(x);
        
        if(mst.isEmpty() || x <= mst.peek()){
            mst.push(x);
        }
    }
    
    public void pop() {
        int x = st.pop();
        if(x == mst.peek()) mst.pop();
    }
    
    public int top() {
        return st.peek();
    }
    
    public int getMin() {
        return mst.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
